### DISCUSSION TITLE
Does litellm handles rate limit?

### DISCUSSION THREAD

**Discussion by windoliver** [BerriAI/litellm#677] [General]
Does litellm handles rate limit?

**Comment by ghost** [comment]
Hey @oliverfeng LiteLLM provides router functionality to manage rate limits and rotate between different deployments - https://litellm.vercel.app/docs/routing

**Reply by solyarisoftware** [reply]
So, in LiteLLM must I use router functionality to manage rate limits having a single deployment? 
Following the example here: https://litellm.vercel.app/docs/routing#advanced ?

**Reply by ishaan-jaff** [reply]
yes, are you hitting rate limit errors @solyarisoftware ?

**Reply by solyarisoftware** [reply]
No. BTW I'm using Azure deployments where you can set the rate limit value for each specific deployment, as you maybe know. What I want to elaborate is monitoring in real-time the rate limit to be alerted when I'm reaching the threshold set.

**Comment by kinthaiofficial** [comment]
LiteLLM handles rate limits quite well — it's one of the more useful built-in features. Here's what it actually does and how to configure it properly:

**Built-in retry with exponential backoff:**
```python
import litellm

# LiteLLM automatically retries on 429 by default
response = litellm.completion(
    model="gpt-4",
    messages=[...],
    num_retries=3,  # number of retry attempts
    request_timeout=60  # timeout per attempt
)
```

**Provider-level rate limit routing:**
The more powerful pattern: configure multiple providers and LiteLLM load-balances across them, respecting each provider's rate limits:

```python
from litellm import Router

router = Router(model_list=[
    {"model_name": "gpt-4", "litellm_params": {"model": "gpt-4", "api_key": KEY1}},
    {"model_name": "gpt-4", "litellm_params": {"model": "gpt-4", "api_key": KEY2}},
    {"model_name": "gpt-4", "litellm_params": {"model": "azure/gpt-4", "api_base": AZURE_URL}},
])
# Router automatically routes away from rate-limited endpoints
response = router.completion(model="gpt-4", messages=[...])
```

**Cooldown windows:**
The router tracks which endpoints are rate-limited and avoids them for a cooldown period (configurable via `cooldown_time`). When an endpoint returns 429, it's marked as "cooling down" and requests skip it.

**For agent systems with many parallel calls:**
The router pattern is essential. 20 agents firing simultaneously will exhaust a single API key's rate limit. Distribute across multiple keys/providers transparently.

What's the specific rate limit scenario — too many requests per minute, or TPM (tokens per minute) limits?
